core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/config/UserStoreConfigXMLProcessor.java

```diff
@@ -430,143 +427,142 @@ private static PrivateKey getPrivateKey() {
         byte[] cipherTextBytes = Base64.decode(propValue.trim());
+        String algorithm = null;
+        byte[] decryptedValue;
+        boolean isInternalKeyStoreEncryptionEnabled = false;
+        boolean isSymmetricKeyEncryptionEnabled = false;
+        ServerConfigurationService config =
+                UserStoreMgtDSComponent.getServerConfigurationService();
+        if (config != null) {
+            String encryptionKeyStore = config.getFirstProperty(ENCRYPTION_KEYSTORE);
 
-        privateKey = (privateKey == null) ? getPrivateKey() : privateKey;
-        if (privateKey == null) {
-            throw new org.wso2.carbon.user.api.UserStoreException(
-                    "Private key initialization failed. Cannot decrypt the userstore password.");
+            if (INTERNAL_KEYSTORE.equalsIgnoreCase(encryptionKeyStore)) {
+                isInternalKeyStoreEncryptionEnabled = true;
+            }
+            String cryptoProvider = config.getFirstProperty(CRYPTO_PROVIDER);
+            if (SYMMETRIC_KEY_CRYPTO_PROVIDER.equalsIgnoreCase(cryptoProvider)) {
+                isSymmetricKeyEncryptionEnabled = true;
+            }
         }
 
-        if(cipherTransformation != null) {
-            // extract the original cipher if custom transformation is used configured in carbon.properties.
-            CipherHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
-            if (cipherHolder != null) {
-                // cipher with meta data.
+        if (isInternalKeyStoreEncryptionEnabled && isSymmetricKeyEncryptionEnabled) {
+
+            throw new CryptoException(String.format("Userstore encryption can not be supported due to " +
+                    "conflicting configurations: '%s' and '%s'. When using internal keystore, assymetric crypto " +
+                    "provider should be used.", INTERNAL_KEYSTORE, SYMMETRIC_KEY_CRYPTO_PROVIDER));
+        } else if (isInternalKeyStoreEncryptionEnabled || isSymmetricKeyEncryptionEnabled) {
+
+            if (cipherTransformation != null) {
+                // extract the original cipher if custom transformation is used configured in carbon.properties.
+                CipherMetaDataHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
+                if (cipherHolder != null) {
+                    // cipher with meta data.
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cipher transformation for decryption : " + cipherHolder.getTransformation());
+                    }
+                    algorithm = cipherHolder.getTransformation();
+                    cipherTextBytes = cipherHolder.getCipherBase64Decoded();
+                } else {
+                    // If the ciphertext is not a self-contained, directly decrypt using transformation configured in
+                    // carbon.properties file
+                    algorithm = cipherTransformation;
+                }
+            }
```

Comment:
If caron.properties is not present can 'cipherTransformation' be null
And if so what happens.

Reply:
Found it

Reply:
yes, cipherTransformation becomes null.
Then the algorithm becomes null.
It will call the cryptoService.decrypt(cipherTextBytes, algorithm, CRYPTO_API_PROVIDER_BC); (line 482)with the null algorithm. But in cryptoservice it will assign the default algorithm if the incoming algorithm is null.

Reply:
+1